src/test/java/streams/part2/exercise/Exercise2.java

```diff
@@ -0,0 +1,242 @@
+                                                   .map(employee -> new Pair<>(employee.getJobHistory().get(0).getEmployer(),
+                                                           employee.getPerson()))
+                                                   .collect(Collectors.groupingBy(Pair::getKey, Collectors.mapping(Pair::getValue, Collectors.toSet())));
+
+        assertThat(result, hasEntry(is("yandex"), contains(employees.get(2).getPerson())));
+        assertThat(result, hasEntry(is("T-Systems"), containsInAnyOrder(employees.get(3).getPerson(), employees.get(5).getPerson())));
+        assertThat(result, hasEntry(is("EPAM"), containsInAnyOrder(
+                employees.get(0).getPerson(),
+                employees.get(1).getPerson(),
+                employees.get(4).getPerson()
+        )));
+    }
+
+    /**
+     * Преобразовать список сотрудников в отображение [компания -> сотрудник, суммарно проработавший в ней наибольшее время].
+     * Гарантируется, что такой сотрудник будет один.
+     */
+    @Test
+    void greatestExperiencePerEmployer() {
+        List<Employee> employees = getEmployees();
+
+        Map<String, Person> collect = employees.stream().flatMap(employee -> employee.getJobHistory()
+                                                                                     .stream()
+                                                                                     .map(jobEntry -> new Trinity(employee.getPerson(),
+                                                                                                                  jobEntry.getEmployer(),
+                                                                                                                  jobEntry.getDuration())))
+                                               .collect(Collectors.toMap(trinity -> new Pair<>(trinity.getPerson(), trinity.getEmployer()),
+                                                                         Trinity::getDuaration,
+                                                                         (t1, t2) -> t1 + t2))
+                                               .entrySet()
+                                               .stream()
+                                               .map(entry -> new Trinity(entry.getKey().getKey(), entry.getKey().getValue(), entry.getValue()))
+                                               .collect(Collectors.toMap(Trinity::getEmployer, t1 -> new Pair<>(t1.getPerson(), t1.getDuaration()),
+                                                       (p1, p2) -> p1.getValue() > p2.getValue() ? p1 : p2))
+                                               .entrySet()
+                                               .stream()
+                                               .map(entry -> new Pair<>(entry.getKey(), entry.getValue().getKey()))
+                                               .collect(Collectors.toMap(Pair::getKey, Pair::getValue));
```

Comment:
You can improve it this way:
```
        Map<String, Person> collect = employees.stream()
                                               .flatMap(employee -> employee.getJobHistory()
                                                                            .stream()
                                                                            .map(jobEntry -> new Trinity(employee.getPerson(),
                                                                                    jobEntry.getEmployer(),
                                                                                    jobEntry.getDuration())))
                                               .collect(toMap(trinity -> new Pair<>(trinity.getPerson(), trinity.getEmployer()),
                                                       Trinity::getDuaration, Integer::sum))
                                               .entrySet()
                                               .stream()
                                               .map(entry -> new Trinity(entry.getKey().getKey(), entry.getKey().getValue(), entry.getValue()))
                                               .collect(groupingBy(Trinity::getEmployer, collectingAndThen(maxBy(comparingInt(Trinity::getDuaration)),
                                                       trinity -> trinity.map(Trinity::getPerson)
                                                                         .orElseThrow(RuntimeException::new))));
```
And please, don't use this default `Pair` class, it's really hard to read and understand constructions like `entry -> new Trinity(entry.getKey().getKey(), entry.getKey().getValue(), entry.getValue())`. Have a look at my solution, for example: https://github.com/elefus/epam-lab-java8-2018-09/pull/366/files.